src/sinks/socket.rs

```diff
@@ -3,19 +3,43 @@ use serde::{Deserialize, Serialize};
 #[cfg(unix)]
 use crate::sinks::util::unix::UnixSinkConfig;
 use crate::{
+    codecs::{
+        encoding::{self, FramingConfig, SerializerConfig},
+        JsonSerializerConfig, NewlineDelimitedEncoder, TextSerializerConfig,
+    },
     config::{DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
     sinks::util::{
-        encode_log, encoding::EncodingConfig, tcp::TcpSinkConfig, udp::UdpSinkConfig, Encoding,
+        encoding::{EncodingConfig, EncodingConfigAdapter, EncodingConfigMigrator},
+        tcp::TcpSinkConfig,
+        udp::UdpSinkConfig,
+        Encoding,
     },
 };
+use bytes::BytesMut;
+use tokio_util::codec::Encoder;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Migrator;
+
+impl EncodingConfigMigrator for Migrator {
+    type Codec = Encoding;
+
+    fn migrate(codec: &Self::Codec) -> (Option<Box<dyn FramingConfig>>, Box<dyn SerializerConfig>) {
+        let framing = None;
+        let serializer = match codec {
+            Encoding::Text => Box::new(TextSerializerConfig::new()) as _,
+            Encoding::Json => Box::new(JsonSerializerConfig::new()) as _,
+        };
+        (framing, serializer)
+    }
+}
 
 #[derive(Deserialize, Serialize, Debug)]
-// TODO: add back when serde-rs/serde#1358 is addressed
-// #[serde(deny_unknown_fields)]
```

Comment:
Just curious, why remove all of these (even the working ones)?

Reply:
They don't work any more since we have two adjacent `#[serde(flatten)]` attributes. E.g. deserializing `TcpSinkConfig` for `mode` will unfortunately reject flattened attributes from `encoding` when `#[serde(flatten)]` is present.